libraries/AP_CANManager/AP_CANManager.cpp

```diff
@@ -417,18 +417,42 @@ bool AP_CANManager::handle_can_forward(mavlink_channel_t chan, const mavlink_com
 /*
   handle a CAN_FRAME packet
  */
-void AP_CANManager::handle_can_frame(const mavlink_message_t &msg) const
+void AP_CANManager::handle_can_frame(const mavlink_message_t &msg)
 {
-    const uint16_t timeout_us = 2000;
+    if (frame_buffer == nullptr) {
+        // allocate frame buffer
+        WITH_SEMAPHORE(_sem);
+        // 20 is good for firmware upload
+        uint8_t buffer_size = 20;
+        while (frame_buffer == nullptr && buffer_size > 0) {
+            // we'd like 20 frames, but will live with less
+            frame_buffer = new ObjectBuffer<BufferFrame>(buffer_size);
```

Comment:
If this method is in a thread then we suddenly have two subsystems both doing this buffer-size-reduction check, and one of them dependent on network traffic.  That's going to lead to random behavior.

Reply:
this code is only on one thread (mavlink receive)

Reply:
Sorry, I was refering to the AP_Logger buffer as the second subsystem here.  They're both on the main thread ATM so there's no problem so far.